File: wpilibj/src/main/java/edu/wpi/first/wpilibj/smartdashboard/proto/MechanismRoot2dProto.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.smartdashboard.proto;

import edu.wpi.first.util.protobuf.Protobuf;
import edu.wpi.first.wpilibj.proto.Mechanism2D.ProtobufMechanismLigament2d;
import edu.wpi.first.wpilibj.proto.Mechanism2D.ProtobufMechanismRoot2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismObject2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
import us.hebi.quickbuf.Descriptors.Descriptor;

public class MechanismRoot2dProto implements Protobuf<MechanismRoot2d, ProtobufMechanismRoot2d> {
  @Override
  public Class<MechanismRoot2d> getTypeClass() {
    return MechanismRoot2d.class;
  }

  @Override
  public Descriptor getDescriptor() {
    return ProtobufMechanismRoot2d.getDescriptor();
  }

  @Override
  public ProtobufMechanismRoot2d createMessage() {
    return ProtobufMechanismRoot2d.newInstance();
  }

  @Override
  public MechanismRoot2d unpack(ProtobufMechanismRoot2d msg) {
    var root = new MechanismRoot2d(msg.getName(), msg.getX(), msg.getY());
    for (ProtobufMechanismLigament2d ligament : msg.getLigaments()) {
      root.append(MechanismLigament2d.proto.unpack(ligament));
    }
    return root;
  }

  @Override
  public void pack(ProtobufMechanismRoot2d msg, MechanismRoot2d value) {
    msg.setX(value.getX());
    msg.setY(value.getY());
    for (MechanismObject2d ligament : value.getObjects().values()) {
      var nestedLigament = MechanismLigament2d.proto.createMessage();
      MechanismLigament2d.proto.pack(nestedLigament, (MechanismLigament2d) ligament);
      msg.addLigaments(nestedLigament);
    }
  }
}

File: wpilibc/src/main/native/cpp/smartdashboard/proto/MechanismRoot2dProto.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/smartdashboard/MechanismRoot2d.h"
#include "frc/smartdashboard/MechanismLigament2d.h"
#include "frc/smartdashboard/proto/MechanismRoot2dProto.h"

#include "mechanism2d.pb.h"

google::protobuf::Message* wpi::Protobuf<frc::MechanismRoot2d>::New(
    google::protobuf::Arena* arena) {
  return google::protobuf::Arena::CreateMessage<
      wpi::proto::ProtobufMechanismRoot2d>(arena);
}

frc::MechanismRoot2d wpi::Protobuf<frc::MechanismRoot2d>::Unpack(
    const google::protobuf::Message& msg) {
  auto m = static_cast<const wpi::proto::ProtobufMechanismRoot2d*>(&msg);
  auto root = frc::MechanismRoot2d{m->name(), m->x(), m->y()};
  for (const auto ligament : m->ligaments()) {
    root.Append<frc::MechanismLigament2d>(ligament.name(), wpi::UnpackProtobuf<frc::MechanismLigament2d>(ligament));
  }
}

void wpi::Protobuf<frc::MechanismRoot2d>::Pack(
    google::protobuf::Message* msg, const frc::MechanismRoot2d& value) {
  auto m = static_cast<wpi::proto::ProtobufMechanismRoot2d*>(msg);
  m->set_name(value.GetName());
  m->set_x(value.GetX());
  m->set_y(value.GetY());
  for (auto&& root : value.GetObjects()) {
    // auto mechanismRoot = ;
    wpi::PackProtobuf<frc::MechanismRoot2d>(m->add_ligaments(), *root.second); 
  }
}
